src/parser/parser.ts

```diff
@@ -196,25 +197,28 @@ export class Parser {
         }
         return null;
     }
-    string(): TokenInfo | null {
-        const tok = this._tokenizer.peek();
+    string(): null | Constant {
```

Comment:
is this right? I thought the parser calls `parsestr`? or is this your way of getting around the whole concat mess I dug myself into 😄

Reply:
I don't think so. I think it's more correct for number. Because this.number gets called in the generated parser (not sure if this.string() is called). They probably each have a helper function at this stage but it probably does return a constant in the end having decided what the Constant kind is and any cleaning.

Reply:
had another look at this - and I think the concat mess is still valid. They get called at different points in the generated parser. I'm guessing concatenate deals with `"foo" "bar"` and then once that's over we get to here to create the Constant node.

Reply:
see 7b77c6d for resolution